mne/time_frequency/tfr.py

```diff
@@ -245,6 +249,239 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+    # Check frequencies
+    if not isinstance(frequencies, (list, np.ndarray)):
+        raise ValueError('frequencies must be an array-like.')
+    frequencies = np.asarray(frequencies, dtype=float)
+    if frequencies.ndim != 1:
+        raise ValueError('frequencies must be of shape n_freqs.')
+
+    # Check sfreq
+    if not isinstance(sfreq, (float, int)):
+        raise ValueError('sfreq must be a float or an int.')
+    sfreq = float(sfreq)
+
+    # Check output
+    allowed_ouput = ('complex', 'power', 'phase',
+                     'avg_power_phaselock', 'avg_power', 'avg_phaselock')
+    if output not in allowed_ouput:
+        raise ValueError("Unknown output type.")
+
+    if method not in ('mtm', 'morlet'):
+        raise ValueError('method must be "morlet" or "mtm"')
+
+    # Default zero_mean = True if mtm else False
+    zero_mean = method == 'mtm' if zero_mean is None else zero_mean
+    if not isinstance(zero_mean, bool):
+        raise ValueError('')
+    frequencies = np.asarray(frequencies)
+
+    # XXX Can we compute single-trial phases with mtm?
+    if (method == 'mtm') and (output == 'phase'):
+        raise NotImplementedError(
+            'This function is not optimized to compute the phase using the '
+            'mtm method. Use np.angle of the complex output instead.')
+
+    # Setup wavelet
+    if method == 'morlet':
+        W = morlet(sfreq, frequencies, n_cycles=n_cycles, zero_mean=zero_mean)
+        Ws = [W]  # to have same dimensionality as the 'mtm' case
+        if time_bandwidth != 4.0:
+            raise ValueError('time_bandwidth only applies to "mtm" method.')
+    elif method == 'mtm':
+        Ws = dpss_wavelet(sfreq, frequencies, n_cycles=n_cycles,
+                          time_bandwidth=time_bandwidth, zero_mean=zero_mean)
+
+    # Check wavelets
+    if len(Ws[0][0]) > epoch_data.shape[2]:
+        raise ValueError('At least one of the wavelets is longer than the '
+                         'signal. Use a longer signal or shorter wavelets.')
+
+    # Initialize output
+    decim = _check_decim(decim)
+    n_freqs = len(frequencies)
+    n_epochs, n_chans, n_times = epoch_data[:, :, decim].shape
+    if output in ('power', 'phase', 'avg_power', 'avg_phaselock'):
+        dtype = np.float
+    elif output in ('complex', 'avg_power_phaselock'):
+        # avg_power_phaselock is stored as power + 1i * phase_lock to keep a
+        # simple dimensionality
+        dtype = np.complex
+
+    if 'avg_' in output:
+        out = np.empty((n_chans, n_freqs, n_times), dtype)
+    else:
+        out = np.empty((n_chans, n_epochs, n_freqs, n_times), dtype)
+
+    # Parallel computation
+    parallel, my_cwt, _ = parallel_func(_time_frequency_loop, n_jobs)
+
+    # Parallization is applied across channels.
+    tfrs = parallel(
+        my_cwt(channel, Ws, output, use_fft, 'same', decim)
+        for channel in epoch_data.transpose(1, 0, 2))
+
+    # FIXME: to avoid overheads we should use np.array_split()
```

Comment:
> If there's a single job, I thought mne.parallel was basically skipping joblib; does it not?

Yes, but depending on the array sizes it can end up being more efficient to triage based on `n_jobs==1` or not. Here I doubt it will matter much in terms of memory as it is currently coded.

> IIUC, you're saying that splitting the data into n_chunks, where each chunk has several channels, then this will be more memory consuming than splitting the data into 306 jobs directly. If so, yes I agree. It is a memory / CPU trade-off.

Yeah, there is a tradeoff. Have you looked at the memory and CPU use for multiple use cases for `n_jobs=1` vs `n_jobs=-1`, both when using direct iteration as here and when using `np.array_split` of the data? Sometimes the pickling overhead is really bad, so the additional memory usage is worth it. It would take some testing to know.

> In any case, I think splitting over channels is a better strategy than over epochs as it was before, because it allows computing single trial or inter-trial metrics.

This could be true, it could also be slower again because of the pickling. Now if there are 100 epochs and 306 channels, you're pickling certain parameters 30,000+ times instead of only `n_jobs` times, which could be brutal. Again, it's worth checking to see how it does. We might end up needing two paths of parallelization, one for multiple trials (or more than `N` trials) and one for single-ish trials.

TL;DR: I see how this could be beneficial in some situations, but without someone doing the benchmarks, it's hard to know how often this is beneficial vs detrimental w/r/t CPU and memory tradeoffs.